src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+          PRECONDITION(can_cast_expr<symbol_exprt>(called_function_pointer));
+          auto const &pointer_symbol = to_symbol_expr(called_function_pointer);
+          auto const restriction_iterator =
+            restrictions.restrictions.find(pointer_symbol.get_identifier());
+          if(restriction_iterator != restrictions.restrictions.end())
+          {
+            auto const &restriction = *restriction_iterator;
+            // if we can, we will replace uses of it by the
+            // this is intentionally a copy because we're just about to change
+            // the instruction this iterator points to
+            auto const original_function_call_instruction = *location;
+            *location = goto_programt::make_assertion(
+              false_exprt{},
+              make_function_pointer_restriction_assertion_source_location(
+                original_function_call_instruction.source_location,
+                restriction));
+            auto const assume_false_location = goto_function_body.insert_after(
+              location,
+              goto_programt::make_assumption(
+                false_exprt{},
+                original_function_call_instruction.source_location));
+            // this is mutable because we'll update this at the end of each
+            // loop iteration to always point at the start of the branch
+            // we created
+            auto else_location = location;
+            auto const end_if_location = goto_function_body.insert_after(
+              assume_false_location, goto_programt::make_skip());
+            for(auto const &restriction_target : restriction.second)
+            {
+              auto new_instruction = original_function_call_instruction;
+              // can't use get_function_call because that'll return a const ref
+              const symbol_exprt &function_pointer_target_symbol_expr =
+                goto_model.symbol_table.lookup_ref(restriction_target)
+                  .symbol_expr();
+              to_code_function_call(new_instruction.code).function() =
+                function_pointer_target_symbol_expr;
+              auto const goto_end_if_location =
+                goto_function_body.insert_before(
+                  else_location,
+                  goto_programt::make_goto(
+                    end_if_location,
+                    original_function_call_instruction.source_location));
+              auto const replaced_instruction_location =
+                goto_function.second.body.insert_before(
+                  goto_end_if_location, new_instruction);
+              else_location = goto_function.second.body.insert_before(
+                replaced_instruction_location,
+                goto_programt::make_goto(
+                  else_location,
+                  notequal_exprt{
+                    pointer_symbol,
+                    address_of_exprt{function_pointer_target_symbol_expr}}));
```

Comment:
Hard to read, suggestions to improve:

 - pull out bodies of loops to reduce indentation
 - use make_range / filter rather than lots of nested ifs
 - invert the ifs with a `continue` to reduce indentation
 - use smaller functions rather than comments to explain blocks of code

Reply:
Inverted the ifs :green_heart: